Add guidance to eBPF extension writers about pointers in contexts (#4408)
microsoft/ebpf-for-windows: docs/eBpfExtensions.md

`sample_program_context_header_t` when invoking the eBPF program. The header is not accessible
by the program.

##### Pointer members in program contexts

The sample program context above contains two pointer fields, `data_start` and `data_end`.
If the intent of an extension is to provide compatibility with some program type that exists on Linux,

> and pointer members exist in the program context, there is an inherent problem.

Author: ```suggestion
and pointer members exist in the program context, there is a potential problem to be aware of.
```

> On Linux, "pointers" in some program contexts are defined as 32-bit integers, even on 64-bit platforms (conversion
> is done at program load time), whereas the example above will result in 64-bit pointers on

Author: ```suggestion
On Linux, "pointers" in some program contexts are defined as 32-bit integers, even on 64-bit platforms (with conversion
done at program load time), whereas the example above will result in 64-bit pointers on
```

64-bit platforms.  eBPF for Windows aims to provide source compatibility, but not binary compatibility.

> The problem often goes unnoticed until verification of a program that works fine on Linux failing

Author: ```suggestion
The issue may go unnoticed until verification of a program that works fine on Linux failing
```

> unexpectedly due to the program hard coding the context structure it expects, which of course won't

Author: ```suggestion
unexpectedly on Windows due to the program hard coding the context structure it expects, which of course won't
```

match what the Windows extension uses, since the context has a different offset for `data_end`.

> As such, even source compatibility is not provided in this sense.

Author: ```suggestion
```

> As such, extensions intended to provide source-compatibility with Linux must minimally document this

Author: ```suggestion
As such, extensions intended to provide source-compatibility with Linux should minimally document this
```

`sample_program_context_header_t` when invoking the eBPF program. The header is not accessible
by the program.

##### Pointer members in program contexts

The sample program context above contains two pointer fields, `data_start` and `data_end`.
If the intent of an extension is to provide compatibility with some program type that exists on Linux,
and pointer members exist in the program context, there is an inherent problem.

> On Linux, pointers in program contexts are always 32-bit, even on 64-bit platforms (conversion

Reviewer: How did you figure this out? BTF works off of the assumption that pointers are 8 bytes. I'd be surprised if this was true in general.

Author: It is probably specific to those two structs (maybe others) but the point is to call out that implementers need to pay attention to this.

Author: @lmb I updated text

> On Linux, pointers in program contexts are always 32-bit, even on 64-bit platforms (conversion

Author: ```suggestion
On Linux, "pointers" in some program contexts are defined as 32-bit integers, even on 64-bit platforms (conversion
```
